File: _dev/desktop/updater.js
/**
 * Site Bundle Updater
 *
 * Checks timetested.bible for a newer APP_VERSION. If found, downloads
 * a site-bundle.tar.gz, extracts it to userData, and serves from the
 * updated bundle on next launch.
 *
 * Storage layout in app.getPath('userData'):
 *   site-bundles/
 *     current-version          (text file containing version number)
 *     v1234567890/             (extracted site bundle)
 */

const { app, net } = require('electron');
const path = require('path');
const fs = require('fs');
const { execFile } = require('child_process');

const SITE_URL = 'https://timetested.bible';
const VERSION_URL = `${SITE_URL}/version.js`;
const BUNDLE_URL = `${SITE_URL}/site-bundle.tar.gz`;
const BUNDLES_DIR = path.join(app.getPath('userData'), 'site-bundles');
const CURRENT_FILE = path.join(BUNDLES_DIR, 'current-version');

/**
 * Get the path to the latest downloaded site bundle, or null if none exists.
 */
function getUpdatedBundlePath() {
  try {
    if (!fs.existsSync(CURRENT_FILE)) return null;
    const version = fs.readFileSync(CURRENT_FILE, 'utf8').trim();
    const bundlePath = path.join(BUNDLES_DIR, `v${version}`);
    if (fs.existsSync(path.join(bundlePath, 'index.html'))) {
      return bundlePath;
    }
  } catch (e) {
    console.error('[Updater] Error reading current bundle:', e.message);
  }
  return null;
}

/**
 * Fetch the remote APP_VERSION from timetested.bible/version.js.
 * The file contains: const APP_VERSION = 1234567890;
 */
async function fetchRemoteVersion() {
  return new Promise((resolve) => {
    const request = net.request(VERSION_URL);
    let body = '';
    request.on('response', (response) => {
      if (response.statusCode !== 200) return resolve(null);
      response.on('data', (chunk) => { body += chunk.toString(); });
      response.on('end', () => {
        const match = body.match(/APP_VERSION\s*=\s*(\d+)/);
        resolve(match ? parseInt(match[1], 10) : null);
      });
    });
    request.on('error', () => resolve(null));
    request.end();
  });
}

/**
 * Get the local APP_VERSION from the currently served bundle.
 */
function getLocalVersion(appPath) {
  try {
    const versionFile = path.join(appPath, 'version.js');
    if (!fs.existsSync(versionFile)) return null;
    const content = fs.readFileSync(versionFile, 'utf8');
    const match = content.match(/APP_VERSION\s*=\s*(\d+)/);
    return match ? parseInt(match[1], 10) : null;
  } catch (e) {
    return null;
  }
}

/**
 * Download a file using Electron's net module.
 * @returns {Promise<Buffer>}
 */
function downloadFile(url) {
  return new Promise((resolve, reject) => {
    const request = net.request(url);
    const chunks = [];
    request.on('response', (response) => {
      if (response.statusCode !== 200) {
        reject(new Error(`HTTP ${response.statusCode}`));
        return;
      }
      response.on('data', (chunk) => chunks.push(chunk));
      response.on('end', () => resolve(Buffer.concat(chunks)));
    });
    request.on('error', reject);
    request.end();
  });
}

/**
 * Extract a .tar.gz buffer to a directory using the system tar command.
 * tar is available on macOS, Linux, and Windows 10+ (bsdtar).
 */
function extractTarGz(tarGzPath, destDir) {
  return new Promise((resolve, reject) => {
    fs.mkdirSync(destDir, { recursive: true });
    execFile('tar', ['xzf', tarGzPath, '-C', destDir], (err) => {
      if (err) reject(err);
      else resolve();
    });
  });
}

/**
 * Download and extract the site bundle.
 */
async function downloadBundle(version) {
  const destDir = path.join(BUNDLES_DIR, `v${version}`);
  const tempTar = path.join(BUNDLES_DIR, 'download.tar.gz');
  const tempDir = path.join(BUNDLES_DIR, 'downloading');

  try {
    fs.mkdirSync(BUNDLES_DIR, { recursive: true });

    // Clean up any previous failed attempts
    for (const p of [tempTar, tempDir]) {
      if (fs.existsSync(p)) fs.rmSync(p, { recursive: true });
    }

    console.log(`[Updater] Downloading site bundle v${version}...`);

    const buffer = await downloadFile(BUNDLE_URL);
    fs.writeFileSync(tempTar, buffer);

    console.log(`[Updater] Downloaded ${(buffer.length / 1024 / 1024).toFixed(1)} MB, extracting...`);

    await extractTarGz(tempTar, tempDir);

    // Verify the download
    if (!fs.existsSync(path.join(tempDir, 'index.html'))) {
      throw new Error('Downloaded bundle missing index.html');
    }

    // Atomic swap: rename temp to versioned dir
    if (fs.existsSync(destDir)) fs.rmSync(destDir, { recursive: true });
    fs.renameSync(tempDir, destDir);

    // Update the current version marker
    fs.writeFileSync(CURRENT_FILE, String(version));
    console.log(`[Updater] Installed site bundle v${version}`);

    fs.unlinkSync(tempTar);

    return true;
  } catch (e) {
    console.error('[Updater] Download failed:', e.message);
    for (const p of [tempTar, tempDir]) {
      try { if (fs.existsSync(p)) fs.rmSync(p, { recursive: true }); } catch (_) {}
    }
    return false;
  }
}

// All downloaded bundles are kept — user can switch between them in Settings.

/**
 * Revert to the bundled version (delete current-version marker).
 * On next launch, getAppPath() falls through to the shipped bundle.
 */
function revertToBuiltin() {
  try {
    if (fs.existsSync(CURRENT_FILE)) fs.unlinkSync(CURRENT_FILE);
    console.log('[Updater] Reverted to built-in bundle');
    return true;
  } catch (e) {
    console.error('[Updater] Revert failed:', e.message);
    return false;
  }
}

/**
 * List available bundle versions (downloaded + built-in).
 * @param {string} builtinPath - path to the shipped bundle
 * @returns {{ version: number, path: string, active: boolean, builtin: boolean }[]}
 */
function listBundles(builtinPath) {
  const bundles = [];
  const activeVersion = (() => {
    try {
      if (fs.existsSync(CURRENT_FILE)) return fs.readFileSync(CURRENT_FILE, 'utf8').trim();
    } catch (e) {}
    return null;
  })();

  // Built-in bundle
  const builtinVersion = getLocalVersion(builtinPath);
  if (builtinVersion) {
    bundles.push({
      version: builtinVersion,
      path: builtinPath,
      active: !activeVersion,
      builtin: true
    });
  }

  // Downloaded bundles
  try {
    if (fs.existsSync(BUNDLES_DIR)) {
      for (const entry of fs.readdirSync(BUNDLES_DIR)) {
        if (!entry.startsWith('v')) continue;
        const fullPath = path.join(BUNDLES_DIR, entry);
        if (!fs.statSync(fullPath).isDirectory()) continue;
        const ver = getLocalVersion(fullPath);
        if (ver) {
          bundles.push({
            version: ver,
            path: fullPath,
            active: activeVersion === String(ver),
            builtin: false
          });
        }
      }
    }
  } catch (e) {}

  return bundles.sort((a, b) => b.version - a.version);
}

/**
 * Switch to a specific downloaded bundle version.
 * @param {number} version
 * @returns {boolean}
 */
function switchToVersion(version) {
  const bundlePath = path.join(BUNDLES_DIR, `v${version}`);
  if (!fs.existsSync(path.join(bundlePath, 'index.html'))) return false;
  try {
    fs.mkdirSync(BUNDLES_DIR, { recursive: true });
    fs.writeFileSync(CURRENT_FILE, String(version));
    console.log(`[Updater] Switched to bundle v${version}`);
    return true;
  } catch (e) {
    console.error('[Updater] Switch failed:', e.message);
    return false;
  }
}

/**
 * Check for updates and download if available. Non-blocking — runs in background.
 * @param {string} appPath - current site bundle path being served
 * @param {function} onUpdateReady - callback(newVersion) when download completes
 */
async function checkForUpdates(appPath, onUpdateReady) {
  try {
    const localVersion = getLocalVersion(appPath);
    if (!localVersion) {
      console.log('[Updater] Could not determine local version, skipping');
      return;
    }

    console.log(`[Updater] Local: ${localVersion}`);
    const remoteVersion = await fetchRemoteVersion();
    if (!remoteVersion) {
      console.log('[Updater] Could not fetch remote version (offline?)');
      return;
    }

    console.log(`[Updater] Remote: ${remoteVersion}`);
    if (remoteVersion <= localVersion) {
      console.log('[Updater] Up to date');
      return;
    }

    console.log(`[Updater] Update available: ${localVersion} → ${remoteVersion}`);
    const ok = await downloadBundle(remoteVersion);
    if (ok && onUpdateReady) onUpdateReady(remoteVersion);
  } catch (e) {
    console.error('[Updater] Check failed:', e.message);
  }
}

module.exports = {
  getUpdatedBundlePath, checkForUpdates, getLocalVersion,
  revertToBuiltin, listBundles, switchToVersion
};

File: _dev/desktop-tauri/src-tauri/src/updater.rs
//! Site-bundle updater — Rust port of `_dev/desktop/updater.js`.
//!
//! Mirrors the Electron app exactly so the two desktop shells behave identically:
//!   - checks https://timetested.bible/version.js for a newer APP_VERSION
//!   - downloads site-bundle.tar.gz, extracts it, serves from it next launch
//!
//! On-disk layout under the app data dir (same scheme as Electron):
//!   site-bundles/
//!     current-version        (text file containing the active version number)
//!     v<version>/            (extracted site bundle; must contain index.html)

use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};

// Content updates ride with GitHub Releases — the `latest/download/<asset>` permalink
// always serves the newest published (non-draft, non-prerelease) release's asset, with no
// API call or rate limits. The desktop-release workflow uploads version.js +
// site-bundle.tar.gz to each release. (ureq follows the 302 to the asset host.)
const RELEASES_URL: &str =
    "https://github.com/timetestedbible/timetestedbible.github.io/releases/latest/download";

fn version_url() -> String {
    format!("{RELEASES_URL}/version.js")
}
fn bundle_url() -> String {
    format!("{RELEASES_URL}/site-bundle.tar.gz")
}

/// `<app_data_dir>/site-bundles`
pub fn bundles_dir(app_data_dir: &Path) -> PathBuf {
    app_data_dir.join("site-bundles")
}
fn current_file(app_data_dir: &Path) -> PathBuf {
    bundles_dir(app_data_dir).join("current-version")
}

/// One entry in the version picker — identical shape to the Electron JSON.
#[derive(Serialize, Clone)]
pub struct Bundle {
    pub version: u64,
    pub path: String,
    pub active: bool,
    pub builtin: bool,
}

/// Path of the latest downloaded bundle, or None to fall back to the built-in one.
pub fn get_updated_bundle_path(app_data_dir: &Path) -> Option<PathBuf> {
    let cur = current_file(app_data_dir);
    let version = fs::read_to_string(&cur).ok()?.trim().to_string();
    if version.is_empty() {
        return None;
    }
    let bundle = bundles_dir(app_data_dir).join(format!("v{version}"));
    if bundle.join("index.html").is_file() {
        Some(bundle)
    } else {
        None
    }
}

/// Parse `const APP_VERSION = 1234567890;` out of a version.js string.
fn parse_version(js: &str) -> Option<u64> {
    let idx = js.find("APP_VERSION")?;
    let after = &js[idx..];
    let eq = after.find('=')?;
    after[eq + 1..]
        .trim_start()
        .chars()
        .take_while(|c| c.is_ascii_digit())
        .collect::<String>()
        .parse()
        .ok()
}

/// The APP_VERSION inside a bundle's version.js (built-in or downloaded).
pub fn get_local_version(bundle_path: &Path) -> Option<u64> {
    let js = fs::read_to_string(bundle_path.join("version.js")).ok()?;
    parse_version(&js)
}

/// The current remote APP_VERSION (None if offline / unreachable).
pub fn fetch_remote_version() -> Option<u64> {
    let body = ureq::get(&version_url()).call().ok()?.into_string().ok()?;
    parse_version(&body)
}

/// Download + extract the bundle for `version`. Returns true on success.
/// Uses pure-Rust gzip+tar (flate2 + tar) — no dependency on a system `tar`.
pub fn download_bundle(app_data_dir: &Path, version: u64) -> bool {
    let dir = bundles_dir(app_data_dir);
    let dest = dir.join(format!("v{version}"));
    let tmp = dir.join("downloading");

    let run = || -> std::io::Result<()> {
        fs::create_dir_all(&dir)?;
        if tmp.exists() {
            fs::remove_dir_all(&tmp)?;
        }
        fs::create_dir_all(&tmp)?;

        // Download to memory then stream-extract.
        let resp = ureq::get(&bundle_url())
            .call()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e.to_string()))?;
        let reader = resp.into_reader();
        let gz = flate2::read::GzDecoder::new(reader);
        let mut archive = tar::Archive::new(gz);
        archive.unpack(&tmp)?;

        // Verify
        if !tmp.join("index.html").is_file() {
            return Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                "downloaded bundle missing index.html",
            ));
        }

        // Atomic-ish swap
        if dest.exists() {
            fs::remove_dir_all(&dest)?;
        }
        fs::rename(&tmp, &dest)?;
        fs::write(current_file(app_data_dir), version.to_string())?;
        Ok(())
    };

    match run() {
        Ok(()) => {
            eprintln!("[Updater] Installed site bundle v{version}");
            true
        }
        Err(e) => {
            eprintln!("[Updater] Download failed: {e}");
            let _ = fs::remove_dir_all(&tmp);
            false
        }
    }
}

/// Delete the current-version marker → next launch serves the built-in bundle.
pub fn revert_to_builtin(app_data_dir: &Path) -> bool {
    let cur = current_file(app_data_dir);
    if cur.exists() {
        if let Err(e) = fs::remove_file(&cur) {
            eprintln!("[Updater] Revert failed: {e}");
            return false;
        }
    }
    true
}

/// List built-in + downloaded bundles, newest first. Mirrors listBundles() in JS.
pub fn list_bundles(app_data_dir: &Path, builtin_path: &Path) -> Vec<Bundle> {
    let active = fs::read_to_string(current_file(app_data_dir))
        .ok()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty());

    let mut out: Vec<Bundle> = Vec::new();

    if let Some(v) = get_local_version(builtin_path) {
        out.push(Bundle {
            version: v,
            path: builtin_path.display().to_string(),
            active: active.is_none(),
            builtin: true,
        });
    }

    if let Ok(entries) = fs::read_dir(bundles_dir(app_data_dir)) {
        for entry in entries.flatten() {
            let p = entry.path();
            if !p.is_dir() {
                continue;
            }
            let name = entry.file_name().to_string_lossy().to_string();
            if !name.starts_with('v') {
                continue;
            }
            if let Some(v) = get_local_version(&p) {
                out.push(Bundle {
                    version: v,
                    path: p.display().to_string(),
                    active: active.as_deref() == Some(v.to_string().as_str()),
                    builtin: false,
                });
            }
        }
    }

    out.sort_by(|a, b| b.version.cmp(&a.version));
    out
}

/// Point the current-version marker at an already-downloaded bundle.
pub fn switch_to_version(app_data_dir: &Path, version: u64) -> bool {
    let bundle = bundles_dir(app_data_dir).join(format!("v{version}"));
    if !bundle.join("index.html").is_file() {
        return false;
    }
    let dir = bundles_dir(app_data_dir);
    if fs::create_dir_all(&dir).is_err() {
        return false;
    }
    fs::write(current_file(app_data_dir), version.to_string()).is_ok()
}

/// Background check: returns Some(new_version) if a newer bundle was downloaded.
pub fn check_for_updates(app_data_dir: &Path, serving_path: &Path) -> Option<u64> {
    let local = get_local_version(serving_path)?;
    let remote = fetch_remote_version()?;
    if remote <= local {
        return None;
    }
    eprintln!("[Updater] Update available: {local} -> {remote}");
    if download_bundle(app_data_dir, remote) {
        Some(remote)
    } else {
        None
    }
}
